### PR TITLE
Users component

### PR DESCRIPTION
**Pull request by IsraaSulaiman**
Relates #19

### CODE REVIEW

File: client/src/App.js [hunk 1]
```diff
@@ -40,7 +40,7 @@ class App extends Component {
               />
               <Route
                 exact
-                path="/project/:projectId/:scrumId"
+                path="/project/:projectId"
```
Comment (ashatat) [inline client/src/App.js]: what this change has to do with the users component?

File: client/src/components/Users/index.jsx [hunk 2]
```diff
@@ -1,9 +1,237 @@
...
+        //change message
+        this.setState({ passwordError: false, saved: true });
+      } else {
+        this.setState({ passwordError: true });
+      }
+    }
+  };
+
+  handleEditUserInfo = (event, record, columnName) => {
+    if (record.id === this.state.users.length) {
```
Comment (ashatat) [inline client/src/components/Users/index.jsx]: You can't rely on `record.id` that has to do anything with the `users` indexes
at some point, it should be different.

File: client/src/components/Users/select/style.css [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  border-radius: 5px;
+  padding: .3rem .5rem;
+  cursor: pointer;
+  border: none;
+  font-size: .9rem;
+  background: white;
+}
+
+.users__row:nth-child(odd) .users__select {
+  background: rgb(213, 238, 241, 0.2);
```
Comment (ashatat) [inline client/src/components/Users/select/style.css]:
```suggestion
  background: rgba(213, 238, 241, 0.2);
```

File: client/src/components/Users/index.jsx [hunk 4]
```diff
@@ -1,9 +1,237 @@
...
+
+  AddPassword = password => {
+    this.setState({ password, show: false });
+  };
+  cancel = () => {
+    this.setState({ show: false })
+  }
+
+  render() {
+    const columns = [
```
Comment (ashatat) [inline client/src/components/Users/index.jsx]: This way each time the `render()` function called you'll declare the columns again which is fine if the columns are dynamic.
In our case, the columns ("the header of the table") are static except the last one which declares the dropDown menu

so I suggest you move the declaration of the columns in the class body without the last column
and then add the last column only inside the render function, something like this :
```jsx
this.columns[3] = {
        render: props => {
          return (
            <Dropdown
              key={props.id}
              trigger={['click']}
              rowId={props.id}
              overlay={
                <UserMenu
                  rowId={props.id}
                  submitPassword={this.handleAddPassword}
                  usersLength={this.state.users.length}
                  handleDeleteUser={this.handleDeleteUser}
                  showPasswordPopup={this.showForm}
                />
              }
            >
              <Icon onClick={() => this.handleRow(props.id)} title="click" className="user__ellipsis" type="ellipsis" />
            </Dropdown>
          );
        },
      };
```

As I said, the way you did it works just fine, but I think it'd be better to declare only what you need to change when the state changes inside the render.

File: client/src/components/Users/helpers.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
+import Swal from 'sweetalert2';
+
+//Filtering Function
+function filter(users) {
+  const usernames = [];
+  const nameOptions = [];
+  users.forEach(user => {
+    const username = user['user_name'];
+    if (usernames.indexOf(username) === -1 && username) {
```
[the remark this answers is not in the document]
Comment (IsraaSulaiman) [reply client/src/components/Users/helpers.js]: ```if (usernames.indexOf(username) === -1 && username)```

Looks like you did not notice that I do check if it is null or not!
 ;)

File: client/src/components/Users/index.jsx (hunk 2 above)
Comment (IsraaSulaiman) [reply client/src/components/Users/index.jsx]: I will change it to the following:
```javascript
const {users} = this.state;
if(record.id === users[users.length].id)```

File: client/src/components/Users/index.jsx (hunk 4 above)
Comment (IsraaSulaiman) [reply client/src/components/Users/index.jsx]: Why do you think the dropdown is the only column that is dynamic? 
Nothing static in this table. Inputs are editable

File: client/src/components/Users/index.jsx [hunk 6]
```diff
@@ -1,9 +1,237 @@
...
+      users: [...users, { id: users.length + 1, user_name: '', full_name: '', role: 'developer' }],
+    });
+  }
+  componentDidUpdate(prevProps, prevState) {
+    const { users } = this.state;
+    if (prevState.users !== users) {
+      const allUsers = users.slice(0);
+      const nameOptions = filter(allUsers);
+      this.setState({ nameOptions });
+    }
```
[the remark this answers is not in the document]
Comment (IsraaSulaiman) [reply client/src/components/Users/index.jsx]: Yes, you're right.
 But if I did not use componentDidUpdate, how can I update the filter options each time the user makes a change to the table cells?

[the remark this answers is not in the document]
Comment (ashatat) [reply client/src/components/Users/index.jsx]: yeah, you are right, I couldn't think of any other logic
Just wanted you to know what is happening

File: client/src/components/Users/index.jsx [hunk 7]
```diff
@@ -1,9 +1,223 @@
...
+          return (
+            <Editable
+              html={value}
+              onChange={event => this.handleEditUserInfo(event, record, 'user_name')}
+              tagName="span"
+              className="users__cell"
+            />
+          );
+        },
+        filters: filter(this.state.users).nameOptions,
```
Comment (IsraaSulaiman) [inline client/src/components/Users/index.jsx]: @ashatat

Comment (ashatat) [reply client/src/components/Users/index.jsx]: :+1:  Good

**Pull request merged**